clustering/consul

```diff
@@ -0,0 +1,463 @@
+    import simplejson as json
+
+try:
+    import consul
+except ImportError, e:
+    print "failed=True msg='python-consul required for this module. "\
+          "see http://python-consul.readthedocs.org/en/latest/#installation'"
+    sys.exit(1)
+
+
+def register_with_consul(module):
+
+    state = module.params.get('state')
+
+    if state == 'present':
+        add(module)
+    else:
+        remove(module)
+
+
+def add(module):
+    ''' adds a service or a check depending on supplied configuration'''
+    check = parse_check(module)
+    service = parse_service(module)
+
+    if not service and not check:
+        module.fail_json(msg='a name and port are required to register a service')
+
+    if service:
+        if check:
+            service.add_check(check)
+        add_service(module, service)
+    elif check:
+        add_check(module, check)
+
+
+def remove(module):
+    ''' removes a service or a check '''
+    service_id = module.params.get('service_id') or module.params.get('service_name')
+    check_id = module.params.get('check_id') or module.params.get('check_name')
+    if not (service_id or check_id):
+        module.fail_json(msg='services and checks are removed by id or name.'\
+                            ' please supply a service id/name or a check id/name')
+    if service_id:
+        remove_service(module, service_id)
+    else:
+        remove_check(module, check_id)
+
+
+def add_check(module, check):
+    ''' registers a check with the given agent. currently there is no way
+    retrieve the full metadata of an existing check  through the consul api.
+    Without this we can't compare to the supplied check and so we must assume
+    a change. '''
+    if not check.name:
+        module.fail_json(msg='a check name is required for a node level check,'\
+                                ' one not attached to a service')
+
+    consul_api = get_consul_api(module)
+    check.register(consul_api)
+
+    module.exit_json(changed=True,
+                     check_id=check.check_id,
+                     check_name=check.name,
+                     script=check.script,
+                     interval=check.interval,
+                     ttl=check.ttl)
+
+
+def remove_check(module, check_id):
+    ''' removes a check using its id '''
+    consul_api = get_consul_api(module)
+
+    if check_id in consul_api.agent.checks():
+        consul_api.agent.check.deregister(check_id)
+        module.exit_json(changed=True, id=check_id)
+
+    module.exit_json(changed=False, id=check_id)
+
+
+def add_service(module, service):
+    ''' registers a service with the the current agent '''
+    result = service
+    changed = False
+
+    consul_api = get_consul_api(module)
+    existing = get_service_by_id(consul_api, service.id)
+
+    # there is no way to retreive the details of checks so if a check is present
+    # in the service it must be reregistered
+    if service.has_checks() or not(existing or existing == service):
```

Comment:
Marin,

The intention here is that the service should only be registered if it is
different from what is already registered. This is so we can report
'changed' correctly You can't retrieve the details for the checks from the
api so there is now way to tell if the checks have changed so you have to
assume so and re-register the service.

 if service.has_checks() or not(existing or existing == service):

This is saying if the user supplies checks or the service provided is not
the same as the existing registered check. The tags are included in the
equals comparison so it should re-register if you change a tag. If you're
seeing different this is definitely an issue. I'll put together a test to
make sure this is working.

On Thu, Feb 19, 2015 at 6:22 PM, Marin <notifications@github.com> wrote:

> In clustering/consul
> <https://github.com/ansible/ansible-modules-extras/pull/225#discussion_r25011042>
> :
>
> > +        module.exit_json(changed=True, id=check_id)
> > +
> > +    module.exit_json(changed=False, id=check_id)
> > +
> > +
> > +def add_service(module, service):
> > +    ''' registers a service with the the current agent '''
> > +    result = service
> > +    changed = False
> > +
> > +    consul_api = get_consul_api(module)
> > +    existing = get_service_by_id(consul_api, service.id)
> > +
> > +    # there is no way to retreive the details of checks so if a check is present
> > +    # in the service it must be reregistered
> > +    if service.has_checks() or not(existing or existing == service):
>
> I am wondering about this - in my use case I always want the service to be
> registered even if there aren't any checks. Could has_checks() be optional
> or make it so the service is always reregistered?
>
> Removing this line works for me (ie if i change tags on a service they are
> properly updated in consul)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ansible/ansible-modules-extras/pull/225/files#r25011042>
> .
>